modular_skyrat/master_files/code/datums/chatmessage.dm

```diff
@@ -0,0 +1,55 @@
+/*
+	The color clamp for the preference selector is set here:
+	'modular_skyrat/master_files/code/modules/client/preferences/runechat.dm'
+*/
+
+////
+//	Modular runechat color selector solution
+/datum/chatmessage/generate_image(text, atom/target, mob/owner, datum/language/language, list/extra_classes, lifespan)
+
+	if(ismob(target))
+		var/mob/actor = target
+
+		if(actor.client?.prefs)
+			if(actor.chat_color_name != actor.name) // So we don't generate each message
+
+				// Check for the custom color if the actor has one
+				var/enable_chat_color_player = actor.client.prefs.read_preference(/datum/preference/toggle/enable_chat_color_player)
+				if (enable_chat_color_player && actor.name == actor.real_name)
+					var/chat_color_player = actor.client.prefs.read_preference(/datum/preference/color/chat_color_player)
+					var/hex_value = sanitize_hexcolor(chat_color_player)
+					target.chat_color = hex_value
+					target.chat_color_darkened = (hex_value + "c0") // Add an alpha-channel to darken
+					target.chat_color_name = actor.name
+
+
+				// If the actor is copying someone else, find out who
+				if(actor.name != actor.real_name && actor.name != "Unknown")
+					for(var/mob/living/carbon/human/copy as anything in GLOB.player_list)
+						if(actor.name == copy.real_name)
```

Comment:
This is quite possibly the most expensive and asinine way to do this.
Is it really necessary to loop through a list of possibly 80 or more players?
An associative list of real name to chat color is most likely a better approach.

Reply:
It doesn't loop, and it also only calls once per name swap. I think your tone is very uncalled for. What is the difference between 'an associative list' and the list of names im searching through with this global call?

Reply:
I honestly can't tell if you are just being snarky, or are actually offering a real alternative for me to find a player in the player list. Because yes, it really is necessary to search a list of possible 80 or more players, how else can I find out their chat color preference? And how big of a performance hit really is it to do that once in a while, its not like players are changing names highly frequently.

Reply:
I'm not trying to be mean, i'm just saying that an associative list is more performant.
As in, a list indexing real name to chat color. This way the preference only has to be read once, and in case someone takes the name, you can easily take the color from the list. I don't know how else to elaborate.